pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+					<AccountCurrencyTransferAllowance<T>>::remove((
+						&account_id,
+						&currency_id,
+						&receiver,
+					));
+					Self::decrement_or_remove_allowance_count(&account_id, &currency_id)?;
+					Self::deposit_event(Event::TransferAllowancePurged {
+						sender_account_id: account_id,
+						currency_id,
+						receiver,
+					});
+					Ok(())
+				}
+				None => Err(DispatchError::from(Error::<T>::NoMatchingAllowance)),
+			}
+		}
+
+		#[pallet::call_index(3)]
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(0, 1).ref_time())]
+		/// Adds an account/currency delay
+		/// Calling on an existing combination will update the existing delay value
+		pub fn add_or_update_allowance_delay(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			delay: T::BlockNumber,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			<AccountCurrencyDelay<T>>::insert(&account_id, &currency_id, &delay);
+			Self::deposit_event(Event::TransferAllowanceDelaySet {
+				sender_account_id: account_id,
+				currency_id,
+				delay,
+			});
+			Ok(())
+		}
+
+		#[pallet::call_index(4)]
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(0, 1).ref_time())]
+		/// Removes an existing sending account/currency delay
+		pub fn remove_allowance_delay(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			match <AccountCurrencyDelay<T>>::get(&account_id, &currency_id) {
+				Some(_) => {
+					<AccountCurrencyDelay<T>>::remove(&account_id, &currency_id);
+					Self::deposit_event(Event::TransferAllowanceDelayRemoval {
+						sender_account_id: account_id,
+						currency_id,
+					});
+					Ok(())
+				}
+				None => {
+					Self::deposit_event(Event::TransferAllowanceDelayRemoval {
+						sender_account_id: account_id,
+						currency_id,
+					});
+					Err(DispatchError::from(Error::<T>::NoMatchingDelay))
+				}
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Increments number of allowances present for a sending account/currency set.
+		/// If no allowances set, an entry with 1 added, if entry already present, it is then incremented.
+		pub fn increment_or_create_allowance_count(
+			account_id: &T::AccountId,
+			currency_id: &T::CurrencyId,
+		) -> DispatchResult {
+			// not using try_mutate here as we're not sure if key exits, and we're already doing a some value check on result of exists query check
+			match <AccountCurrencyTransferRestriction<T>>::get(account_id, currency_id) {
+				Some(allowance_count) if allowance_count > 0 => {
+					let new_allowance_count = allowance_count
+						.checked_add(1)
+						.ok_or(Error::<T>::AllowanceCountOverflow)?;
+					<AccountCurrencyTransferRestriction<T>>::insert(
+						account_id,
+						currency_id,
+						new_allowance_count,
+					);
+					Ok(())
+				}
+				Some(_) => Err(DispatchError::Other("Impossible allowance count")),
```

Comment:
In which scenario would this branch be possible?

Reply:
I just had a comment in the same line. This case doesn't belong to the model and can never happen.

Reply:
I've combined the storage items for delay and count as per @mikiquantum's suggestion, which has removed that case as a possibility with the types so that branch is no longer there. :)
(was not possible to reach before, however types would have accommodated that branch)